### DISCUSSION TITLE
P2P remote REPL across internet (WebRTC)

### DISCUSSION THREAD

**Discussion by vshymanskyy** [micropython/micropython#15329] [Show and tell]
While working on Web based MicroPython IDE, I came across this idea of using WebRTC to create a remote connection to the device.

It was rather easy to create a proof of concept. It works like this:

> `MPY Board`  <---USB---> `Bridge Web Page` <---Internet---> `ViperIDE`

Demo: https://viper-ide.org/bridge.html

![image](https://github.com/micropython/micropython/assets/1506708/0c76bf79-1a47-493e-84e8-cba40e1f2177)

Also, this raises an interesting question:
**What if MicroPython could run the WebRTC client directly?**
This is different from WebREPL because enables connection from anywhere. Alternatively I can run a WebSocket relay server, and modify WebREPL to act as a Client instead of Server.

If anyone has experience with this and would like to participate - you're welcome!

**Comment by andrewleech** [comment]
I'd never heard of WebRTC however a quick search suggests it's aimed at "real-time audio, video streaming and video conferencing applications". 
What are you using it for? I would generally use basic websockets for real time comms from server direct to client.

Regardless, in terms of communicating directly with a device, at work I regularly use a design pattern where we use stm32 configured with two CDC serial ports. The first is normal repl, the second is treated like a fake socket. On device I use a simple wrapper function to pass this serial port stream to microdot so we can make http requests directly over this serial port. 

On the PC side, custom apps either use a similarly wrapped python http library to make requests over the serial port rather than opening new sockets, or we can run a simple script that proxies the serial port to a local socket to then use a web browser or any other app.

There's some slight complexities in terms of needing a lock on the PC side so that only one request can run at a time but it works surprisingly well! 

Because it's a single stream however, there's no way to keep a second persistent connection open like a websocket.

... Alternatively, LWIP does have support for PPP so it should be possible to get this running now on devices, even without a native network interface (https://github.com/micropython/micropython/pull/12872) which then might make it easier to get PPP running over the USB/serial to get a "real" network connection. That would mean using a modem driver of the host to configure the connection however. 
Similarly, with LWIP it might be possible to use the new dynamic usb runtime configuration to get the USB working as a "USB Network Adaptor" to have a real network connection that way. Then the host will need some other way to configure and identify the right IP address to talk to the device. 

Either way, if you can get a real network connection then protocols like WebRTC or websockets should just work.

**Comment by vshymanskyy** [comment]
WebRTC allows direct connection between ViperIDE and the bridge app, over the internet.
It doesn't usually require an aditional websocket/relay server in the middle (only a signalling server).

Currently it requires a bridge app that converts Serial to WebRTC

**Comment by vshymanskyy** [comment]
After some testing, i figured out that WebRTC is rather cumbersome, and I'll probably need to run a TURN server to make it working reliably. Given that getting WebRTC running on the board directly requires significant additional efforts, I'll focus on a different approach:

> `MicroPython Device`  --- websocket ---> `viper-ide.org` <----- websocket --- `ViperIDE client (browser)`

Basically this is the same as `WebREPL` modified to work via a **Secure WebSocket relay server**.
It should be much more practical and easy to achieve at this stage.

I'll probably do some prototyping for this in the following weeks. Will keep you posted 😎

**Reply by Josverl** [reply]
Likely both STUN and TURN as well as some fallback proxies will be needed to get around the any different combinations of SNATS and DNATS.
Either that - or a overlay network such as Tailscale

**Reply by vshymanskyy** [reply]
Using a WebSocket relay should be reliable and simple, at the expense of increased latency and (possibly) increased load on the server. However if we're talking about ~1000 simultaneous REPL sessions, I don't think it makes any sense to dive into WebRTC.

**Comment by vshymanskyy** [comment]
The initial version of a remote **Secure WebSocket Relay** connection is now available for public testing:
https://github.com/vshymanskyy/ViperIDE/blob/main/docs/Web-REPL-Relay.md
